Spark Walmart Data Analysis Project

Let's get some quick practice with your new Spark DataFrame skills, you will be asked some basic questions about some stock market data, in this case Walmart Stock from the years 2012-2017. This exercise will just ask a bunch of questions, unlike the machine learning exercises, which will be a little looser and be in the form of "Consulting Projects".

For now, just answer the questions and complete the tasks below.

Use the walmart_stock.csv file to Answer and complete the tasks below!



1.Load the Walmart Stock CSV File, have Spark infer the data types.

In [76]:
# Import the pandas library as pd
import pandas as pd  

# Read the CSV file 'walmart_stock.csv' into a pandas DataFrame
dataframe = pd.read_csv('walmart_stock.csv')   

# Print the entire DataFrame to the console
print(dataframe)

# Get and print the data types of each column in the DataFrame
datatype = dataframe.dtypes
print(datatype)


            Date       Open       High        Low      Close    Volume  \
0     2012-01-03  59.970001  61.060001  59.869999  60.330002  12668800   
1     2012-01-04  60.209999  60.349998  59.470001  59.709999   9593300   
2     2012-01-05  59.349998  59.619999  58.369999  59.419998  12768200   
3     2012-01-06  59.419998  59.450001  58.869999  59.000000   8069400   
4     2012-01-09  59.029999  59.549999  58.919998  59.180000   6679300   
...          ...        ...        ...        ...        ...       ...   
1253  2016-12-23  69.430000  69.750000  69.360001  69.540001   4803900   
1254  2016-12-27  69.300003  69.820000  69.250000  69.699997   4435700   
1255  2016-12-28  69.940002  70.000000  69.260002  69.309998   4875700   
1256  2016-12-29  69.209999  69.519997  69.120003  69.260002   4298400   
1257  2016-12-30  69.120003  69.430000  68.830002  69.120003   6889500   

      Adj Close  
0     52.619235  
1     52.078475  
2     51.825539  
3     51.459220  
4     51.616215  
...

2.What are the column names?
A.['Date', 'Open', 'High', 'Low', 'Close', 'Volume', 'Adj Close']

In [77]:
# Get the column names
column_names = dataframe.columns

# Print or use the column names as needed
print("Column Names:")
print(column_names)


Column Names:
Index(['Date', 'Open', 'High', 'Low', 'Close', 'Volume', 'Adj Close'], dtype='object')


3.What does the Schema look like?
A.root
 |-- Date: timestamp (nullable = true)
 |-- Open: double (nullable = true)
 |-- High: double (nullable = true)
 |-- Low: double (nullable = true)
 |-- Close: double (nullable = true)
 |-- Volume: integer (nullable = true)
 |-- Adj Close: double (nullable = true)

In [78]:
dataframe.info()  #getting all information about the dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1258 entries, 0 to 1257
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Date       1258 non-null   object 
 1   Open       1258 non-null   float64
 2   High       1258 non-null   float64
 3   Low        1258 non-null   float64
 4   Close      1258 non-null   float64
 5   Volume     1258 non-null   int64  
 6   Adj Close  1258 non-null   float64
dtypes: float64(5), int64(1), object(1)
memory usage: 68.9+ KB


4.Print out the first 5 columns.

In [79]:
dataframe.head(5)  #printing first 5 data from dataframe

Date       Open       High        Low      Close    Volume  Adj Close
0  2012-01-03  59.970001  61.060001  59.869999  60.330002  12668800  52.619235
1  2012-01-04  60.209999  60.349998  59.470001  59.709999   9593300  52.078475
2  2012-01-05  59.349998  59.619999  58.369999  59.419998  12768200  51.825539
3  2012-01-06  59.419998  59.450001  58.869999  59.000000   8069400  51.459220
4  2012-01-09  59.029999  59.549999  58.919998  59.180000   6679300  51.616215

5.Use describe() to learn about the DataFrame.
A.+-------+------------------+-----------------+-----------------+-----------------+-----------------+-----------------+
|summary|              Open|             High|              Low|            Close|           Volume|        Adj Close|
+-------+------------------+-----------------+-----------------+-----------------+-----------------+-----------------+
|  count|              1258|             1258|             1258|             1258|             1258|             1258|
|   mean| 72.35785375357709|72.83938807631165| 71.9186009594594|72.38844998012726|8222093.481717011|67.23883848728146|
| stddev|  6.76809024470826|6.768186808159218|6.744075756255496|6.756859163732991|  4519780.8431556|6.722609449996857|
|    min|56.389998999999996|        57.060001|        56.299999|        56.419998|          2094900|        50.363689|
|    max|         90.800003|        90.970001|            89.25|        90.470001|         80898100|84.91421600000001|

In [80]:
dataframe.describe() # Generate describe for the DataFrame

Open         High          Low        Close        Volume  \
count  1258.000000  1258.000000  1258.000000  1258.000000  1.258000e+03   
mean     72.357854    72.839388    71.918601    72.388450  8.222093e+06   
std       6.768090     6.768187     6.744076     6.756859  4.519781e+06   
min      56.389999    57.060001    56.299999    56.419998  2.094900e+06   
25%      68.627503    69.059998    68.162503    68.632497  5.791100e+06   
50%      73.235000    73.725002    72.839996    73.265000  7.093500e+06   
75%      76.629997    77.094999    76.250000    76.709999  9.394675e+06   
max      90.800003    90.970001    89.250000    90.470001  8.089810e+07   

         Adj Close  
count  1258.000000  
mean     67.238838  
std       6.722609  
min      50.363689  
25%      63.778335  
50%      68.541162  
75%      71.105668  
max      84.914216

6.There are too many decimal places for mean and stddev in the describe() dataframe. Format the numbers to just show up to two decimal places. Pay careful attention to the datatypes that .describe() returns, we didn't cover how to do this exact formatting, but we covered something very similar.

A.+-------+--------+--------+--------+--------+----------+
|summary|    Open|    High|     Low|   Close|    Volume|
+-------+--------+--------+--------+--------+----------+
|  count|1,258.00|1,258.00|1,258.00|1,258.00|     1,258|
|   mean|   72.36|   72.84|   71.92|   72.39| 8,222,093|
| stddev|    6.77|    6.77|    6.74|    6.76| 4,519,780|
|    min|   56.39|   57.06|   56.30|   56.42| 2,094,900|
|    max|   90.80|   90.97|   89.25|   90.47|80,898,100|
+-------+--------+--------+--------+--------+----------+

In [81]:
# Generate describe for the DataFrame
data = dataframe.describe()

# Round the values to two decimal places
rounded_data = data.round(2)
print(rounded_data)


          Open     High      Low    Close       Volume  Adj Close
count  1258.00  1258.00  1258.00  1258.00      1258.00    1258.00
mean     72.36    72.84    71.92    72.39   8222093.48      67.24
std       6.77     6.77     6.74     6.76   4519780.84       6.72
min      56.39    57.06    56.30    56.42   2094900.00      50.36
25%      68.63    69.06    68.16    68.63   5791100.00      63.78
50%      73.24    73.73    72.84    73.26   7093500.00      68.54
75%      76.63    77.09    76.25    76.71   9394675.00      71.11
max      90.80    90.97    89.25    90.47  80898100.00      84.91


7.Create a new dataframe with a column called HV Ratio that is the ratio of the High Price versus volume of stock traded for a day.

A.+--------------------+
|            HV Ratio|
+--------------------+
|4.819714653321546E-6|
|6.290848613094555E-6|
|4.669412994783916E-6|
|7.367338463826307E-6|
|8.915604778943901E-6|
|8.644477436914568E-6|
|9.351828421515645E-6|
| 8.29141562102703E-6|
|7.712212102001476E-6|
|7.071764823529412E-6|
|1.015495466386981E-5|
|6.576354146362592...|
| 5.90145296180676E-6|
|8.547679455011844E-6|
|8.420709512685392E-6|
|1.041448341728929...|
|8.316075414862431E-6|
|9.721183814992126E-6|
|8.029436027707578E-6|
|6.307432259386365E-6|
+--------------------+
only showing top 20 rows

In [82]:
# function to calculate the High/Volume ratio
def hv_ratio(row):
    return row['High'] / row['Volume']

dataframe['HV Ratio'] = dataframe.apply(hv_ratio, axis=1)

# Extract the 'HV Ratio' column from the DataFrame
new_dataframe = dataframe['HV Ratio']
print(new_dataframe.head(20))


0     0.000005
1     0.000006
2     0.000005
3     0.000007
4     0.000009
5     0.000009
6     0.000009
7     0.000008
8     0.000008
9     0.000007
10    0.000010
11    0.000007
12    0.000006
13    0.000009
14    0.000008
15    0.000010
16    0.000008
17    0.000010
18    0.000008
19    0.000006
Name: HV Ratio, dtype: float64


8.What day had the Peak High in Price?
A.datetime.datetime(2015, 1, 13, 0, 0)

In [83]:
# Finding the maximum value in the 'High' column
maximum_high = dataframe['High'].max()

# Creating a new DataFrame containing rows where 'High' is equal to the maximum value
new_dataframe = dataframe[dataframe['High'] == maximum_high]

print(new_dataframe['Date'])


761    2015-01-13
Name: Date, dtype: object


9.What is the max and min of the Volume column?
A.+-----------+-----------+
|max(Volume)|min(Volume)|
+-----------+-----------+
|   80898100|    2094900|
+-----------+-----------+

In [84]:
# Find the maximum volume in the 'Volume' column
max_volume = dataframe['Volume'].max()

# Find the minimum volume in the 'Volume' column
min_volume = dataframe['Volume'].min()

# Print the maximum and minimum volume values
print("Max of volume is:", max_volume)
print("Min of volume is:", min_volume)


Max of volume is: 80898100
Min of volume is: 2094900


10.How many days was the Close lower than 60 dollars?
A.81

In [85]:
data=(dataframe['Close']<60)
print(data.sum())   # Count the number of True values in the boolean mask, indicating 'Close' values less than 60

81


11.What percentage of the time was the High greater than 80 dollars ?
A. 9.141494

In [100]:
# Create a boolean mask where 'High' values are greater than 80
data=(dataframe['High']>80)
greater_than_80=data.sum()
total_count=data.count()

# Calculate the percentage of 'High' values that are greater than 80
print(100*greater_than_80/total_count)

9.141494435612083


Spark Walmart Data Analysis Project

Let's get some quick practice with your new Spark DataFrame skills, you will be asked some basic questions about some stock market data, in this case Walmart Stock from the years 2012-2017. This exercise will just ask a bunch of questions, unlike the machine learning exercises, which will be a little looser and be in the form of "Consulting Projects".

For now, just answer the questions and complete the tasks below.

Use the walmart_stock.csv file to Answer and complete the tasks below!



1.Load the Walmart Stock CSV File, have Spark infer the data types.

In [ ]:
# Import the pandas library as pd
import pandas as pd  

# Read the CSV file 'walmart_stock.csv' into a pandas DataFrame
dataframe = pd.read_csv('walmart_stock.csv')   

# Print the entire DataFrame to the console
print(dataframe)

# Get and print the data types of each column in the DataFrame
datatype = dataframe.dtypes
print(datatype)


            Date       Open       High        Low      Close    Volume  \
0     2012-01-03  59.970001  61.060001  59.869999  60.330002  12668800   
1     2012-01-04  60.209999  60.349998  59.470001  59.709999   9593300   
2     2012-01-05  59.349998  59.619999  58.369999  59.419998  12768200   
3     2012-01-06  59.419998  59.450001  58.869999  59.000000   8069400   
4     2012-01-09  59.029999  59.549999  58.919998  59.180000   6679300   
...          ...        ...        ...        ...        ...       ...   
1253  2016-12-23  69.430000  69.750000  69.360001  69.540001   4803900   
1254  2016-12-27  69.300003  69.820000  69.250000  69.699997   4435700   
1255  2016-12-28  69.940002  70.000000  69.260002  69.309998   4875700   
1256  2016-12-29  69.209999  69.519997  69.120003  69.260002   4298400   
1257  2016-12-30  69.120003  69.430000  68.830002  69.120003   6889500   

      Adj Close  
0     52.619235  
1     52.078475  
2     51.825539  
3     51.459220  
4     51.616215  
...

12.What is the Pearson correlation between High and Volume?
A.+-------------------+
| corr(High, Volume)|
+-------------------+
|-0.3384326061737161|
+-------------------+

In [87]:
# Calculate the Pearson correlation between 'High' and 'Volume'
correlation = dataframe['High'].corr(dataframe['Volume'])

# Display the result
print("Pearson correlation between High and Volume:")
print(correlation)

Pearson correlation between High and Volume:
-0.3384326061737167


13.What is the max High per year?
A.+----+---------+
|Year|max(High)|
+----+---------+
|2015|90.970001|
|2013|81.370003|
|2014|88.089996|
|2012|77.599998|
|2016|75.190002|
+----+---------+

In [99]:
# Define the group_by_year function to extract the year from the datetime object
def group_by_year(dataframe):
    return dataframe['Date'].split('-')[0]

dataframe['Years']=dataframe.apply(group_by_year,axis=1)

# Group the DataFrame by the 'Years' column
ans=dataframe.groupby('Years')
# Calculate the maximum High price for each year
print(ans['High'].max())

Years
2012    77.599998
2013    81.370003
2014    88.089996
2015    90.970001
2016    75.190002
Name: High, dtype: float64


14.What is the average Close for each Calendar Month?
In other words, across all the years, what is the average Close price for Jan,Feb, Mar, etc... Your result will have a value for each of these months.

A.+-----+-----------------+
|Month|       avg(Close)|
+-----+-----------------+
|    1|71.44801958415842|
|    2|  71.306804443299|
|    3|71.77794377570092|
|    4|72.97361900952382|
|    5|72.30971688679247|
|    6| 72.4953774245283|
|    7|74.43971943925233|
|    8|73.02981855454546|
|    9|72.18411785294116|
|   10|71.57854545454543|
|   11| 72.1110893069307|
|   12|72.84792478301885|
+-----+-----------------+

In [98]:
# Define the group_by_month function to extract the month from the datetime object
def group_by_month(dataframe):
    return dataframe['Date'].split('-')[1]

dataframe['Month']=dataframe.apply(group_by_month,axis=1)

# Group the DataFrame by the 'Month' column
ans=dataframe.groupby('Month')

# Calculate the mean closing price for each month
print(ans['Close'].mean())

Month
01    71.448020
02    71.306804
03    71.777944
04    72.973619
05    72.309717
06    72.495377
07    74.439719
08    73.029819
09    72.184118
10    71.578545
11    72.111089
12    72.847925
Name: Close, dtype: float64
